daemons/ipa-sam/ipa_sam.c

```diff
@@ -1974,46 +1981,111 @@ static bool handle_cross_realm_princs(struct ipasam_private *ipasam_state,
+		r_tdo_alias == NULL) {
 		ok = false;
 		goto done;
 	}
 
 	switch (mod) {
 		case SET_PRINC:
-			/* Create Kerberos principal for inbound trust, enabled by default */
-			ok   = set_krb_princ(ipasam_state, tmp_ctx, princ_r, NULL, pwd, trusted_dn, KRB_PRINC_CREATE_DEFAULT);
-			/* Create Kerberos principal corresponding to TDO in AD for SSSD usage, disabled by default */
-			ok |= set_krb_princ(ipasam_state, tmp_ctx, princ_tdo, saltprinc_tdo, pwd, trusted_dn,
-					    KRB_PRINC_CREATE_DISABLED | KRB_PRINC_CREATE_AGENT_PERMISSION);
-			if ((trust_direction & LSA_TRUST_DIRECTION_OUTBOUND) != 0) {
-				/* Create Kerberos principal for outbound trust, enabled by default */
-				ok |= set_krb_princ(ipasam_state, tmp_ctx, princ_l, NULL, pwd, trusted_dn, KRB_PRINC_CREATE_DEFAULT);
+			/* We must use two sets by two principals here because
+			 * they are used for different needs and must have
+			 * different salts */
+
+			failed = 0;
+			/* INBOUND TRUST */
+			if ((trust_direction & LSA_TRUST_DIRECTION_INBOUND) != 0) {
+				/* First: krbtgt/<OUR REALM>@<REMOTE REALM>, enabled by default
+				 * in case of the inboud trust */
+				failed += !set_krb_princ(ipasam_state, tmp_ctx, princ_r, NULL,
+							 pwd_outgoing, trusted_dn,
+							 KRB_PRINC_CREATE_DEFAULT);
+
+				/* Second: <OUR FLATNAME$>@<REMOTE REALM> is only used
+				 * for SSSD to be able to talk to AD DCs but it has to
+				 * have canonical name set to <OUR FLATNAME>$ because
+				 * this is the salt used by AD DCs when using this
+				 * principal, otherwise authentication will fail.
+				 *
+				 * *disable* use of this principal on our side as it is
+				 * only used to retrieve trusted domain credentials by
+				 * AD Trust Agents across the IPA topology */
+				failed += !set_krb_princ(ipasam_state, tmp_ctx,
+							 r_tdo_alias, princ_r_tdo,
+							 pwd_incoming, trusted_dn,
+							 (KRB_PRINC_CREATE_DISABLED |
+							  KRB_PRINC_CREATE_AGENT_PERMISSION));
+
+	                        ok = (failed == 0);
+				if (!ok) {
+					goto done;
+				}
 			}
-			if (!ok) {
-				goto done;
+
+			failed = 0;
+			/* OUTBOUND TRUST */
+			if ((trust_direction & LSA_TRUST_DIRECTION_OUTBOUND) != 0) {
+				/* First: krbtgt/<REMOTE REALM>@<OUR REALM>, enabled by default */
+				failed += !set_krb_princ(ipasam_state, tmp_ctx,
+							 princ_l, NULL,
+							 pwd_outgoing, trusted_dn,
+							 KRB_PRINC_CREATE_DEFAULT);
+
+				/* Second: <REMOTE FLAT NAME>$@<OUR REALM>, enabled by default
+				 * as it is used for a remote DC to authenticate against IPA Samba
+				 *
+				 * A local account for the outbound trust must have
+				 * POSIX and SMB identities associated with our domain but we associate
+				 * them with the trust domain object itself */
+				failed += !set_krb_princ(ipasam_state, tmp_ctx,
+							 princ_l_tdo, l_tdo_alias,
+							 pwd_incoming, trusted_dn,
+							 KRB_PRINC_CREATE_DEFAULT);
+
+	                        ok = (failed == 0);
```

Comment:
The indention of this line looks wrong.

Reply:
Please ignore it. I have a patch that unifies indentation everywhere in ipasam but it also breaks git history as the patch is touching almost every line of code.